### ISSUE TITLE
[BUG] maxTries and tryTimeoutInSeconds are not applied expectedly when using RequestRetryOptions.

### ISSUE THREAD

**Issue by takmnagaya** [Azure/azure-sdk-for-java#20729]
**Describe the bug**

maxTries and tryTimeoutInSeconds are not applied expectedly when using RequestRetryOptions.

tryTimeoutInSeconds was set to 1 second and maxTries was set to 5. Then I tried downloading a large file (100MB).

The large file cannot be download in 1 second. So retry will be executed 5 times.
But retry was executed only once and then download was succeeded unexpectedly.

**Could you please tell me if this issue is due to client application?**
**Could you please tell me if there are any workarounds?**

RequestRetryOptions was configured as follows.
```java
        BlobServiceClient blobServiceClient = new BlobServiceClientBuilder()
        .connectionString(connectStr)
        .httpLogOptions(new HttpLogOptions().setLogLevel(HttpLogDetailLevel.BASIC))
        .retryOptions(new RequestRetryOptions(RetryPolicyType.EXPONENTIAL, 5, 1, new Long(10000), new Long(20000), null))
        .buildClient();
```

RequestRetryOptions 
https://docs.microsoft.com/en-us/java/api/com.azure.storage.common.policy.requestretryoptions.requestretryoptions?view=azure-java-stable#com_azure_storage_common_policy_RequestRetryOptions_RequestRetryOptions_com_azure_storage_common_policy_RetryPolicyType_java_lang_Integer_java_lang_Integer_java_lang_Long_java_lang_Long_java_lang_String_

***Exception or Stack Trace***

I can see that a request was timed out at around 1:51:03.
After 10 seconds later, retry was executed at 1:51:14. But after that no retry was executed.
```
4月 16, 2021 1:51:03 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:14 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
```

Here is the full stdout.
```
Picked up _JAVA_OPTIONS: -Xmx512m
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------< com.blobs.quickstart:blob-quickstart-v12-javav2 >-----------
[INFO] Building blob-quickstart-v12-javav2 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:java (default-cli) @ blob-quickstart-v12-javav2 ---
4月 16, 2021 1:51:03 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:14 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (701 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (694 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (692 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (699 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (693 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (700 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (700 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:15 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (700 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:16 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:16 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (202 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (274 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (182 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (201 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (202 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (190 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (205 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:17 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (199 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:18 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:18 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:18 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (184 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:18 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (187 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:18 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (212 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:18 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:18 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (196 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:18 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:19 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (200 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:19 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:19 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (185 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:19 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:19 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (210 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:20 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:20 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (225 ms, 4194304-byte body)<-- END HTTP
4月 16, 2021 1:51:20 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:20 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (180 ms, 2516582-byte body)<-- END HTTP
4月 16, 2021 1:51:28 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 16, 2021 1:51:28 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (217 ms, 4194304-byte body)<-- END HTTP
[WARNING] thread Thread[Thread-1,5,com.blobs.quickstart.App] was interrupted but is still alive after waiting at least 14999msecs
[WARNING] thread Thread[Thread-1,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-2,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-3,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-4,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-5,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-6,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-7,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-8,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-9,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[elastic-evictor-1,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption        
[WARNING] thread Thread[elastic-2,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-1,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[reactor-http-nio-1,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption       
[WARNING] thread Thread[reactor-http-nio-2,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[reactor-http-nio-3,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption       
[WARNING] thread Thread[reactor-http-nio-4,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption       
[WARNING] thread Thread[reactor-http-nio-5,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption       
[WARNING] thread Thread[reactor-http-nio-6,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[reactor-http-nio-7,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption       
[WARNING] thread Thread[reactor-http-nio-8,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption       
[WARNING] thread Thread[parallel-2,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-3,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-4,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-5,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[elastic-3,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[elastic-4,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-6,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[elastic-5,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-7,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-8,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[elastic-6,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[elastic-7,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[elastic-8,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[elastic-9,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] NOTE: 34 thread(s) did not finish despite being asked to  via interruption. This is not a problem with exec:java, it is a problem with the running code. Although not serious, it should be remedied.
[WARNING] Couldn't destroy threadgroup org.codehaus.mojo.exec.ExecJavaMojo$IsolatedThreadGroup[name=com.blobs.quickstart.App,maxpri=10]
java.lang.IllegalThreadStateException
    at java.lang.ThreadGroup.destroy (ThreadGroup.java:795)
    at org.codehaus.mojo.exec.ExecJavaMojo.execute (ExecJavaMojo.java:293)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)        
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:78)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:567)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  44.081 s
[INFO] Finished at: 2021-04-16T13:51:45+09:00
[INFO] ------------------------------------------------------------------------
```

**To Reproduce**

1. I followed the following tutorial. The version of azure-storage-blob was set to 12.10.2.
https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-java?tabs=powershell
2. Upload large files (100MB) to Storage Account.
3. Navigate to the /src/main/java/com/blobs/quickstart directory
4. Open the App.java file in your editor and change to following code snippet. (Please change file name, Storage name and the access key.)
5. Execute following commands.

```
mvn compile
mvn package
mvn exec:java -D"exec.mainClass"="com.blobs.quickstart.App"
```

***Code Snippet***

```java
/* sample code  */
// https://docs.microsoft.com/ja-jp/azure/storage/blobs/storage-quickstart-blobs-java?tabs=powershell

package com.blobs.quickstart;

/**
 * Azure blob storage v12 SDK quickstart
 */
import com.azure.storage.blob.*;
import java.io.IOException;
////////////////
import com.azure.storage.common.policy.RequestRetryOptions;
import com.azure.storage.common.policy.RetryPolicyType;
import com.azure.core.http.policy.HttpLogOptions;
import com.azure.core.http.policy.HttpLogDetailLevel;
////////////////

import java.io.File;

public class App {
	// static
	static String connectStr = "DefaultEndpointsProtocol=https;AccountName=storagename;AccountKey=accountkey;EndpointSuffix=core.windows.net";
	static String file = "C:\\Users\\username\\Downloads\\testfile";
        static String fileName = "testfile";
        static String containerName = "sample";

	public static void main(String[] args) throws IOException {
        
        BlobServiceClient blobServiceClient = new BlobServiceClientBuilder()
        .connectionString(connectStr)
        .httpLogOptions(new HttpLogOptions().setLogLevel(HttpLogDetailLevel.BASIC))
        .retryOptions(new RequestRetryOptions(RetryPolicyType.EXPONENTIAL, 5, 1, new Long(10000), new Long(20000), null))
        .buildClient();
        
        BlobContainerClient containerClient = blobServiceClient.getBlobContainerClient(containerName);
        BlobClient blobClient = containerClient.getBlobClient(fileName);
        ///////////////////////////
        blobClient.downloadToFile(file);
        ///////////////////////////

        // File delFile;
        // /*
        //  * delete files for test
        //  */
        // delFile = new File(file);
        // delFile.delete();
	}
}

```

**Expected behavior**
Retry should be executed 5 times.

**Screenshots**

**Setup (please complete the following information):**
 - OS: [e.g. iOS] Windows
 - IDE : [e.g. IntelliJ] Visual Studio Code
 - Version of the Library used: azure-storage-blob 12.10.2

FYI
~~~
mvn -v
Picked up _JAVA_OPTIONS: -Xmx512m
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: C:\Users\username\apache-maven-3.6.3\bin\..
Java version: 16, vendor: Oracle Corporation, runtime: C:\Program Files\Java\jdk-16
Default locale: ja_JP, platform encoding: MS932
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
~~~

**Additional context**
This issue was reported from my customer.
Customer would like to fix this issue.

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

**Comment by gapra-msft** [comment]
Hi @takmnagaya 

Thank you for reporting this issue. I will take a look at what may be going wrong.

**Comment by gapra-msft** [comment]
Hi @takmnagaya 

I think I understand what is happening. The RequestRetryOptions apply to individual requests sent to the service, not the API. As you can tell from the logs, the downloadToFile API on the SDK chunks requests to around 4MB by default (so the timeout of 1s is applied to single 4MB chunk download requests, from your logs it looks like every request is taking <1s). 

One way you can see the RequestRetryOptions work successfully is using the downloadToFileWithResponse API and setting ParallelTransferOptions.blockSizeLong to a larger value.

Hope this helps!

**Comment by takmnagaya** [comment]
Hello @gapra-msft 
I really appreciate your help!

I changed chunk size to 400MB and tried downloading a file (1GB) with ParallelTransferOptions.blockSizeLong.
But retry was executed only once...

**Could you please check if this is expected behavior...?**
If you need more information, please feel free to contact me!

I used ParallelTransferOptions.blockSizeLong as follows.

```java
        BlobServiceClient blobServiceClient = new BlobServiceClientBuilder()
        .connectionString(connectStr)
        .httpLogOptions(new HttpLogOptions().setLogLevel(HttpLogDetailLevel.BASIC))
        .retryOptions(new RequestRetryOptions(RetryPolicyType.EXPONENTIAL, 5, 1, new Long(10000), new Long(20000), null))
        .buildClient();
        
        BlobContainerClient containerClient = blobServiceClient.getBlobContainerClient(containerName);
        BlobClient blobClient = containerClient.getBlobClient(fileName);
        ///////////////////////////
        Duration timeout = Duration.ofSeconds(10000);
        // blobClient.downloadToFileWithResponse(new BlobDownloadToFileOptions(file).setDownloadRetryOptions(new DownloadRetryOptions().setMaxRetryRequests(5)), timeout, Context.NONE);
        blobClient.downloadToFileWithResponse(file, null, new ParallelTransferOptions().setBlockSizeLong(400L * 1024 * 1024), null, null, false, null, null, Context.NONE);
        ///////////////////////////
```

I can see that chunk size was changed successfully from logs.
~~~
<-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (719 ms, 419430400-byte body)<-- END HTTP
~~~

First retry was executed at around 4:30:34. But after that, it seemed to me that no retry were executed.
Finally the large file was downloaded successfully. But it was not expected...
Expected behavior is that retries are executed 5 times and download fails.
~~~
4月 20, 2021 4:30:23 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 20, 2021 4:30:34 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 20, 2021 4:30:35 午後 com.azure.core.util.logging.ClientLogger info
~~~

Here are full logs.
~~~
Picked up _JAVA_OPTIONS: -Xmx512m
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------< com.blobs.quickstart:blob-quickstart-v12-javav2 >-----------
[INFO] Building blob-quickstart-v12-javav2 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:java (default-cli) @ blob-quickstart-v12-javav2 ---
4月 20, 2021 4:30:23 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 20, 2021 4:30:34 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 20, 2021 4:30:35 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (754 ms, 419430400-byte body)<-- END HTTP
4月 20, 2021 4:30:35 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 20, 2021 4:30:35 午後 com.azure.core.util.logging.ClientLogger info
情報: --> GET https://yourstoragename.blob.core.windows.net/sample/testfile
4月 20, 2021 4:30:35 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (719 ms, 419430400-byte body)<-- END HTTP
4月 20, 2021 4:30:35 午後 com.azure.core.util.logging.ClientLogger info
情報: <-- 206 https://yourstoragename.blob.core.windows.net/sample/testfile (725 ms, 234881024-byte body)<-- END HTTP
[WARNING] thread Thread[Thread-1,5,com.blobs.quickstart.App] was interrupted but is still alive after waiting at least 15000msecs
[WARNING] thread Thread[Thread-1,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-2,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-3,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-4,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-5,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-6,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-7,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-8,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[Thread-9,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[elastic-evictor-1,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption        
[WARNING] thread Thread[parallel-1,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[reactor-http-nio-1,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[reactor-http-nio-2,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption       
[WARNING] thread Thread[reactor-http-nio-3,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption       
[WARNING] thread Thread[reactor-http-nio-4,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[reactor-http-nio-5,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption       
[WARNING] thread Thread[reactor-http-nio-6,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption       
[WARNING] thread Thread[reactor-http-nio-7,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[reactor-http-nio-8,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption       
[WARNING] thread Thread[parallel-2,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-3,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-4,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-5,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-6,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-7,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] thread Thread[parallel-8,5,com.blobs.quickstart.App] will linger despite being asked to die via interruption
[WARNING] NOTE: 26 thread(s) did not finish despite being asked to  via interruption. This is not a problem with exec:java, it is a problem with the running code. Although not serious, it should be remedied.
[WARNING] Couldn't destroy threadgroup org.codehaus.mojo.exec.ExecJavaMojo$IsolatedThreadGroup[name=com.blobs.quickstart.App,maxpri=10]
java.lang.IllegalThreadStateException
    at java.lang.ThreadGroup.destroy (ThreadGroup.java:795)
    at org.codehaus.mojo.exec.ExecJavaMojo.execute (ExecJavaMojo.java:293)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:78)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:567)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:21 min
[INFO] Finished at: 2021-04-20T16:35:42+09:00
[INFO] ------------------------------------------------------------------------
~~~

**Comment by gapra-msft** [comment]
Hi @takmnagaya Could you change the value to something more dramatic? It looks like some retries are happening but eventually a request is coming back in < 1s, which is causing the retries to not happen 5 times. 

Here is a short PoC that the retry policy does indeed work as intended. 
Basically I mock the http client to return a response after 2 seconds (and I am using the same retry policy options that you shared). If you run this, you will see the retry happen 5 times (and it will ultimately throw an exception when the 5 tries are up)
```
BlobServiceClient serviceClient = new BlobServiceClientBuilder()
                .connectionString(connectionString)
                .httpClient(new HttpClient() {
                    @Override
                    public Mono<HttpResponse> send(HttpRequest httpRequest) {
                        HttpResponse httpResponse = new HttpResponse(httpRequest) {
                            @Override
                            public int getStatusCode() {
                                return 200;
                            }

                            @Override
                            public String getHeaderValue(String s) {
                                return null;
                            }

                            @Override
                            public HttpHeaders getHeaders() {
                                return null;
                            }

                            @Override
                            public Flux<ByteBuffer> getBody() {
                                return null;
                            }

                            @Override
                            public Mono<byte[]> getBodyAsByteArray() {
                                return null;
                            }

                            @Override
                            public Mono<String> getBodyAsString() {
                                return null;
                            }

                            @Override
                            public Mono<String> getBodyAsString(Charset charset) {
                                return null;
                            }
                        };
                        return Mono.just(httpResponse).delaySubscription(Duration.ofSeconds(2));
                    }
                })
                .httpLogOptions(new HttpLogOptions().setLogLevel(HttpLogDetailLevel.BASIC))
                .retryOptions(new RequestRetryOptions(RetryPolicyType.EXPONENTIAL, 5, 1, new Long(10000), new Long(20000), null))
                .buildClient();


        serviceClient.getHttpPipeline().send(new HttpRequest(HttpMethod.HEAD, new URL("https://account.blob.core.windows.net"), new HttpHeaders().put("Content-Length", "0"), Flux.empty()))
                .block();
```

**Comment by gapra-msft** [comment]
Hi @takmnagaya Do you have any further questions? If so, feel free to open a new issue. I will close this one as I believe it has been resolved.
